src/config/createForm.controller.js

```diff
@@ -31,6 +29,25 @@ export default ({ form, formConfirmation }) => {
   });
 
   form.on("submitDone", () => {
-    if (formConfirmation) window.location.href = formConfirmation;
+    pushDataLayer({
```

Comment:
You could make some extra helper functions for extracting details.
Do you not need the category also on this push layer? 

Also is 200ms long enough for the ga360/ga4 to complete on old phones?

Reply:
you might want to hook into submission instead of the 'done' so it is kicked off at the same time as the ajax call back to formio api.

Reply:
seems the data structure of each ga event is pretty different, so i'd reather construct the data manually instead of using a helper. the data structure is based on steve's suggestion.

the GA call just need to be fired and don't need to be waiting for completion, could add a bit more to 500ms for GA tracker listening to the change on window.datalayer

could add abother event "submitButton" for submission request firing, will ask steve for what GA event suitable for it.

Reply:
https://github.com/formio/formio.js/blob/ab210814dcf12f12bc7da7517223fdeae638789f/app/examples/customendpoint.md

hook into 'submit' for the ga tracking
// Triggered when they click the submit button.
    form.on('submit', function(submission) {
    
  then you don't need to do as much waiting on submitdone for the datapush.

Reply:
i think it's better to hook to `submitButton`(triggered when click on submit button) and `submitDone` separately, and have separate GA events
as we'll have a clear picture of how the submission goes, for example:
submitButton -> submitDone = a sucessful submission
submitButton -> submitError = a fail submission

Reply:
you can submit without hitting the button, which is why i'm suggesting the 'submit' action hook. (i.e. in food pantry where there is a hidden submit when you go to the final page plus another if you want to email yourself.

Reply:
cool will use `submit` instead of `submitButton` then

Reply:
Here's the 'submit' logic,
https://github.com/formio/formio.js/blob/ab210814dcf12f12bc7da7517223fdeae638789f/src/Webform.js#L1292

Reply:
after some analysis with steve and testing, here's some findings.
- "submit" event doesn't fire on submission request triggering, it only fires on submission response is success.
- on a single form submission, either "submitDone" or "submitError" will be fired upon an api success/error response.

so we recon that hooking GA event `formio-submission` to `submitDone` and hooking GA event `ngErrorEvent` to `submitError` will capture all form submission events, no matter how the submission is fired (by button or JS).

Reply:
is ngErrorEvent for angular errors? or has this become the 'generic' spa error catcher?